Added `to_observation_cta` method to `DataStoreObservation`

dcfidalgo opened the pull request:
This PR is a next step towards a working `ObservationCTA` class.

I added an example docstring of how to create an `ObservationCTA` from "scratch", that is without an `DataStore` object for CTA's 1DC data.
A new method in `DataStoreObservation` was added to return an equivalent observation as `ObservationCTA` object.

This is still not ready to merge, i'm still missing a test for the new method.
Hopefully i manage to provide a test before Saturday, otherwise i will do it when i'm back from holidays in two weeks.

A next PR would be to subclass `DataStoreObservation(ObservationCTA)` to avoid code duplication.

cdeil commented on gammapy/data/observations.py:
```diff
@@ -359,6 +382,31 @@ def check_observation(self):
+
+        """
+        # maps the ObservationCTA class attributes to the DataStoreObservation properties
+        obs_cta_kwargs = {'obs_id': 'obs_id', 'gti': 'gti', 'events': 'events', 'aeff': 'aeff', 'edisp': 'edisp',
```
Looks like you have a dict where key val are always the same. Use a list instead, no?

dcfidalgo replied:
Yes, i started with a list, but then i had to create a kwargs dict for the return in the end, and i thought like this it would be nicer. It also gives us a bit more flexibility if in some crazy universe in the future the properties and attributes don't match anymore.

cdeil replied:
Hm, OK. I"m not a big fan of the getattr processing here. But I also don't have an idea at the moment how to avoid it if we want to try and load and copy optional things. So OK to keep as-is for now.

cdeil commented on gammapy/data/observations.py:
```diff
@@ -359,6 +382,31 @@ def check_observation(self):
 
+    def to_observation_cta(self):
+        """
+        Convert a `~gammapy.data.DataStoreObservation` to an `~gammapy.data.ObservationCTA`
```
Just put this: "Convert to `~gammapy.data.ObservationCTA`".
It's clear that it's converting from a DataStoreObservation, since this is a method on that class.
(best to not repeat class names too much, because it's harder to maintain when they change)

dcfidalgo replied:
Done

cdeil commented on gammapy/data/observations.py:
```diff
@@ -68,6 +68,29 @@ class ObservationCTA(object):
+    >>>               edisp = EnergyDispersion2D.read(cal_file, hdu='Energy Dispersion'),
+    >>>               bkg = Background3D.read(cal_file),
+    >>>               pointing_radec = EventList.read(data_file).pointing_radec,
+    >>>               observation_live_time_duration = EventList.read(data_file).observation_live_time_duration,
```
You shouldn't read the event list multiple times.
Suggest to read one thing at a time, create local variables and put the call at the end.

Like this:
```
events = EventList.read(...)
...
obs = ObservationCTA(
    events=events,
    ...
    observation_live_time_duration=events.observation_live_time_duration,
)
```

dcfidalgo replied:
Done

cdeil commented on gammapy/data/observations.py:
```diff
@@ -68,6 +68,29 @@ class ObservationCTA(object):
+    >>>               observation_dead_time_fraction = EventList.read(data_file).observation_dead_time_fraction,
+    >>>          )
+    >>>
+    >>> obs = ObservationCTA(obs_id=110380, **kwargs)
```
Also read the OBS_ID from the events header, because hard-coding it is error-prone, no?

dcfidalgo replied:
Done

cdeil commented on gammapy/data/observations.py:
```diff
@@ -68,6 +68,29 @@ class ObservationCTA(object):
+    --------
+    Create an observation from the CTA 1DC:
+
+    >>> from gammapy.data import ObservationCTA, EventList, GTI
```
Maintaining the example here is tedious for you (with all the >>>) and also for me (can't run it easily).
Suggest to put a Python script in `examples/observation_cta.py` for now, and then once this is useful for users it turns into a notebook.
If you want discoverability from here, you can for now mention the example script from the docstring.
Note that we sometimes link to high-level docs or notebooks from class docstrings, if the exmaple is more than 5 lines.

dcfidalgo replied:
Done

cdeil commented on gammapy/data/tests/test_observations.py:
```diff
@@ -37,6 +37,25 @@ def test_data_store_observation(data_store):
+@requires_data('gammapy-extra')
+def test_data_store_observation_to_observation_cta(data_store):
+    from gammapy.data import EventList, GTI, ObservationCTA
+    from gammapy.irf import EffectiveAreaTable2D, EnergyDispersion2D, EnergyDependentMultiGaussPSF
```
Move imports to top of file, use relative imports within gammapy

dcfidalgo replied:
Done

cdeil commented on examples/example_observation_cta.py:
```diff
@@ -0,0 +1,24 @@
+              observation_dead_time_fraction=event_list.observation_dead_time_fraction,
+              )
+
+obs = ObservationCTA(**kwargs)
```
Why make a dict and past it in like that?
Suggest to move the args into the constructor call directly.

dcfidalgo replied:
Done

dcfidalgo commented:
@cdeil  Thanks for the review!
I added a test for the `DataStoreObservation.to_observation_cta` method and implemented @cdeil comments, mainly moving the example to its own file `examples/example_observation_cta.py`.

> Suggest to put a Python script in examples/observation_cta.py for now, and then once this is useful for users it turns into a notebook.

Yes, i think the final goal would be to provide a notebook in the end.

cdeil commented:
@dcfidalgo - Thanks!

I'll merge this in now.

cdeil commented:
Some code formatting in 4e548762 . Merging.

cdeil commented:
> A next PR would be to subclass DataStoreObservation(ObservationCTA) to avoid code duplication.

I would suggest this:

```
class ObservationBase(object):
    # put methods here that are the same. Maybe __str__
    # or some derived properties

class DataStoreObservation(ObservationBase):
    # Same as current class, except a few methods moved to the base class

class ObservationCTA(ObservationBase):
    # Same as current class, except, maybe, not sure, __str__ moved to the base class
```

I.e. use the base class just as "mix-in" to share a few methods. I don't think directly making `ObservationCTA` or `DataStoreObservation` the parent or child is possible, because one is a disk-based obs and the other an in-memory obs, i.e. they are "parallel", not "above" or "under" the other.

Not sure that makes sense. It's great how you go in small steps now and make small PRs that can go in quickly. If possible, please continue like that when you come back to this work.
